encoders/src/main/scala/org/apache/spark/sql/collection/SharedUtils.scala

```diff
@@ -0,0 +1,188 @@
+          ") can't be more than " + MAX_HASH_CAPACITY + " elements or negative")
+    }
+  }
+
+  def taskMemoryManager(context: TaskContext): TaskMemoryManager = context.taskMemoryManager()
+
+  def toUnsafeRow(buffer: ByteBuffer, numColumns: Int): UnsafeRow = {
+    if (buffer eq null) return null
+    val row = new UnsafeRow(numColumns)
+    if (buffer.isDirect) {
+      row.pointTo(null, UnsafeHolder.getDirectBufferAddress(buffer) +
+          buffer.position(), buffer.remaining())
+    } else {
+      row.pointTo(buffer.array(), Platform.BYTE_ARRAY_OFFSET +
+          buffer.arrayOffset() + buffer.position(), buffer.remaining())
+    }
+    row
+  }
+
+  def getAllExecutorsMemoryStatus(
+      sc: SparkContext): Map[BlockManagerId, (Long, Long)] = {
+    val memoryStatus = sc.env.blockManager.master.getMemoryStatus
+    // no filtering for local backend
+    if (isLoner(sc)) memoryStatus else memoryStatus.filter(!_._1.isDriver)
+  }
+
+  def getHostExecutorId(blockId: BlockManagerId): String =
+    TaskLocation.executorLocationTag + blockId.host + '_' + blockId.executorId
+
+  def createStatsBuffer(statsData: Array[Byte], allocator: BufferAllocator): ByteBuffer = {
+    // need to create a copy since underlying Array[Byte] can be re-used
+    val statsLen = statsData.length
+    val statsBuffer = allocator.allocateForStorage(statsLen)
+    statsBuffer.put(statsData, 0, statsLen)
+    statsBuffer.rewind()
+    statsBuffer
+  }
+
+  final def isLoner(sc: SparkContext): Boolean =
+    (sc ne null) && sc.schedulerBackend.isInstanceOf[LocalSchedulerBackend]
+
+  def acquireStorageMemory(objectName: String, numBytes: Long,
+      buffer: UMMMemoryTracker, shouldEvict: Boolean, offHeap: Boolean): Boolean = {
+    val mode = if (offHeap) MemoryMode.OFF_HEAP else MemoryMode.ON_HEAP
+    if (numBytes > 0) {
+      return MemoryManagerCallback.memoryManager.acquireStorageMemoryForObject(objectName,
+        MemoryManagerCallback.storageBlockId, numBytes, mode, buffer, shouldEvict)
+    } else if (numBytes < 0) {
+      MemoryManagerCallback.memoryManager.releaseStorageMemoryForObject(
+        objectName, -numBytes, mode)
+    }
+    true
+  }
+
+  def releaseStorageMemory(objectName: String, numBytes: Long,
+      offHeap: Boolean): Unit = {
+    val mode = if (offHeap) MemoryMode.OFF_HEAP else MemoryMode.ON_HEAP
+    MemoryManagerCallback.memoryManager.
+        releaseStorageMemoryForObject(objectName, numBytes, mode)
+  }
+
+  /** for testing only (a long convoluted name chosen deliberately) */
+  var TEST_RANDOM_BUCKETID_ASSIGNMENT: Boolean = java.lang.Boolean.getBoolean(
+    "SNAPPYTEST_RANDOM_BUCKETID_TO_PARTITION_ASSIGNMENT")
+
+  /**
+   * TODO: check SmartConnectorRDDHelper for implementation
+   * NOTE: JavaDoc for DataReaderFactory.preferredLocations says
+   * that preferredLocations should return host name, so currently
+   * this returns true
+   */
+  def preferHostName(): Boolean = {
+    true
+  }
```

Comment:
The idea here was to use the same format as configured on Spark executors else the preferredLocations gets ignored regardless of what the javadoc of DataReaderFactory says. This code should be same as that in SmartConnectorHelper (change the argument to SparkContext instead of session since that is all which is required).

Reply:
Temporary removed this piece of code from this PR. As it is specifically required in V2 Connector. Will be using this reference while opening the V2Connector related PR.